lib/spack/spack/test/traverse.py

```diff
@@ -284,3 +329,87 @@ def test_breadth_first_versus_depth_first_printing(abstract_specs_chain):
     ^chain-d
 """
     assert s.tree(depth_first=False, cover="edges", **args) == bfs_tree_edges
+
+
+@pytest.fixture()
+def abstract_specs_toposort():
+    # Create a graph that both BFS and DFS would not traverse in topo order, assuming
+    # edges are ordered by target node name. Roots are {A, E} in forward order and
+    # {F, G} in backward order.
+    # forward: DFS([A, E]) traverses [A, B, F, G, C, D, E] (not topo since C < B)
+    # forward: BFS([A, E]) traverses [A, E, B, C, D, F, G] (not topo since C < B)
+    # reverse: DFS([F, G]) traverses [F, B, A, D, C, E, G] (not topo since D < A)
+    # reverse: BFS([F, G]) traverses [F, G, B, A, D, C, E] (not topo since D < A)
+    # E
+    # | A
+    # | | \
+    # | C |
+    # \ | |
+    #   D |
+    #   | /
+    #   B
+    #  / \
+    # F   G
+    return create_dag(
+        nodes=["A", "B", "C", "D", "E", "F", "G"],
+        edges=(
+            ("A", "B", "all"),
+            ("A", "C", "all"),
+            ("B", "F", "all"),
+            ("B", "G", "all"),
+            ("C", "D", "all"),
+            ("D", "B", "all"),
+            ("E", "D", "all"),
+        ),
+    )
+
+
+def test_traverse_topo_nodes(abstract_specs_toposort):
+    nodes = abstract_specs_toposort
+
+    def test_topo(roots, direction="children"):
+        # Ensure the invariant that all parents of specs[i] are in specs[0:i]
+        specs = list(
+            traverse.traverse_nodes(roots, order="topo", cover="nodes", direction=direction)
+        )
+        reverse = "parents" if direction == "children" else "children"
+        for i in range(len(specs)):
+            parents = specs[i].traverse(cover="nodes", direction=reverse, root=False)
+            assert set(list(parents)).issubset(set(specs[:i]))
+
+    # Traverse forward from roots A and E and a non-root D
+    test_topo([nodes["D"], nodes["E"], nodes["A"]], direction="children")
```

Comment:
They shouldn't be called roots if they have in-degree != 0 :man_shrugging:. In `traverse_nodes` the positional argument is called `specs` not `roots`. It's just a bit unfortunate we have a kwarg named `root=True`, and it's always used to mean "skip the traversal starting point"; maybe `skip_start=True/False` was better. I guess `specs` is more like an ordered list of children of an artificial root (and that root induces the sub-dag of interest), so even then the input specs are not roots :p.

Notice though that in most use cases before we did something like this:

```python
visited = set()
for s in specs:
  for t in s.traverse(root=False, visited=visited):
    # do things
```

which is now equivalent to

```python
for t in traverse(specs, root=False)
```

So, the "potential extra complexity" was already an issue before

Reply:
So, I think I can clarify that elsewhere, but maybe in a separate PR. For now I've just updated the relevant comment in the test.

Reply:
Aargh, I realize I *did* use the word `roots` in the test. Will fix that too.

Reply:
> So, the "potential extra complexity" was already an issue before.

I wouldn't consider that the same kind of complexity: I was worried about semantics within a single traversal. Multiple traversals might add other complications.

> That complexity is gone when using order=breadth, since then you know for a fact that non-root input specs are discovered at depth 0.

To clarify, my specific example was not about a non-root input spec, but rather a spec which is a child of two input specs, appearing at different depths WRT each. Nevertheless, I think I see how BFS produces a well-defined (but meaningless) value for the depth of B when `specs=[A, D]`. However, since we allow `specs=[A, D]` for topo,  it should also make sense for that, or it should be forbidden (the latter being my preference). 

Overall, I'm noticing a trend where allowing `specs` to contain nodes that are children of other nodes in `specs` is making it harder to reason about the semantics; I think forbidding this is appealing because I don't know of any use case that requires it (and supporting it seems to add complexity).

> They shouldn't be called roots if they have in-degree != 0

I'm fine with not calling them roots (although don't really see why it's an issue to call them that) although preferably something more descriptive than `specs`.

> In traverse_nodes the positional argument is called specs not roots.
> Aargh, I realize I did use the word roots in the test. Will fix that too.

And in the docstring for traverse_nodes itself:

`specs (list): List of root specs (considered to be depth 0)`

they may not be roots in a strict sense, but the term "root" captures what we intend to do with it in a single (and short) word.

Reply:
I don't really get the point you're making. In environments you may want to pass a list of env roots to topo traverse, in uninstall you wanna pass a list of specs matching a db query, etc. You don't know if one is a child of the other, that requires traversal. Forbidding it is only annoying

Reply:
> in uninstall you wanna pass a list of specs matching a db query

That is an excellent point (i.e. some specs might be transitive children of others). It's kind of a headache thinking about the meaning of multi-root traversals for the different traversal methods, but given this example it is certainly worthwhile for topological ordering.